pkg/apis/gcpfirewall/v1alpha1/types.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Protocol defines network protocols supported for GCP firewall.
+type Protocol string
+
+// CIDR defines a IP block.
+// TODO(sugangli) Modify the validation to include IPv6 CIDRs with FW 3.0 support.
+// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(/(3[0-2]|2[0-9]|1[0-9]|[0-9]))?$`
+type CIDR string
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:shortName=gf
+
+// GCPFirewall describes a GCP firewall spec that can be used to configure GCE
+// firewalls. A GCPFirewallSpec will correspond 1:1 with a GCE firewall rule.
+type GCPFirewall struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired configuration for GCP firewall
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Spec GCPFirewallSpec `json:"spec,omitempty"`
+
+	// Status is the runtime status of this GCP firewall
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Status GCPFirewallStatus `json:"status,omitempty"`
+}
+
+// Action defines the rule action of the firewall rule.
+type Action string
+
+const (
+	// ActionAllow is the Allow Action of GCP Firewall Rule
+	ActionAllow Action = "ALLOW"
+	// ActionDeny is the Deny Action of GCP Firewall Rule. For now, only Allow is supported.
+	ActionDeny Action = "DENY"
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolICMP is the ICMP protocol.
+	ProtocolICMP Protocol = "ICMP"
+)
+
+// GCPFirewallSpec provides the specification of a GCPFirewall.
+// The firewall rule apply to the cluster associated targets (network tags or
+// secure tags) which are deduced by the controller. As a result, the specified
+// rule applies to ALL nodes and pods in the cluster.
+type GCPFirewallSpec struct {
+	// Rule action of the firewall rule. Only allow action is supported. If not
+	// specified, defaults to ALLOW.
+	// +optional
+	// +kubebuilder:validation:Enum=ALLOW
+	// +kubebuilder:default=ALLOW
+	Action Action `json:"action"`
+
+	// If set to true, the GCPFirewall is not synced by the controller.
```

Comment:
nit: If the firewall rule is previously synced, it would be orphanded?

Reply:
from a consumer stand point. If I mark the CR as disabled, the Firewall controller would not act on the Firewall. Then it is effectively orphanded, right? Then whether the consumer (ingress-gce) decides to configure the firewall on its own and whether to GC the CR. It is orthogonal, no?

Also, in the case of FW 3.0. The FW rule will no longer implement this CR once marked disabled. In that case, it would not be orphanded, it would be removed. 

Am I understanding correctly?